src/main/java/org/terasology/launcher/packages/PackageManager.java

```diff
@@ -94,44 +97,62 @@ public void sync() {
-        final Path sourcesFile = launcherDir.resolve(SOURCES_FILENAME);
+        sourcesFile = launcherDir.resolve(SOURCES_FILENAME);
+    }
 
-        // Copy default sources file if necessary
+    public void validateSources() {
         if (Files.notExists(sourcesFile)) {
-            logger.info("Sources file not found. Copying default file to {}", sourcesFile);
-            try {
-                Files.copy(getClass().getResourceAsStream(SOURCES_FILENAME), sourcesFile);
-            } catch (IOException e) {
-                logger.error("Failed to copy default sources file to {}", sourcesFile);
-            }
-        } else {
-            validateSchema(sourcesFile);
+            logger.warn("sources.json not found: {}", sourcesFile);
+            throw new PackageManagerException("sources.json is missing");
         }
+        if (!validateSchema(sourcesFile)) {
+            throw new PackageManagerException("There were errors reading sources.json");
+        }
+    }
 
+    public void copyDefaultSources() {
+        logger.info("Copying default sources file to {}", sourcesFile);
+        try {
+            Files.copy(getClass().getResourceAsStream(SOURCES_FILENAME),
+                    sourcesFile, StandardCopyOption.REPLACE_EXISTING);
+        } catch (IOException e) {
+            logger.error("Failed to copy default sources file to {}", sourcesFile);
+            throw new RuntimeException("Default sources file could not be copied to " + sourcesFile);
+        }
+    }
+
+    public void initDatabase() {
         database = new PackageDatabase(
                 sourcesFile,
-                launcherDir.resolve(DATABASE_FILENAME),
+                sourcesFile.resolveSibling(DATABASE_FILENAME),
                 installDir
         );
     }
 
-    private void validateSchema(final Path jsonFile) {
+    private boolean validateSchema(final Path jsonFile) {
         try (
                 InputStream schemaIn = getClass().getResourceAsStream(SOURCES_SCHEMA);
                 InputStream jsonIn = Files.newInputStream(jsonFile)
         ) {
-            final Schema schema = SchemaLoader.load(new JSONObject(new JSONTokener(schemaIn)));
+            final Schema schema = SchemaLoader.builder()
+                    .schemaClient(SchemaClient.classPathAwareClient())
+                    .resolutionScope("classpath://org/terasology/launcher/packages/schema.json")
+                    .schemaJson(new JSONObject(new JSONTokener(schemaIn)))
+                    .build().load().build();
+
             schema.validate(new JSONArray((new JSONTokener(jsonIn))));
+            return true;
         } catch (ValidationException e) {
             logger.error("sources.json has invalid value at: {}", e.getPointerToViolation());
-            throw new RuntimeException("Failed to validate sources.json.");
+        } catch (JSONException e) {
+            logger.error("sources.json has syntax error: {}", e.getMessage());
```

Comment:
This doesn't seem to work properly yet. If I add a trailing comma to an entry in the default souces.json, the launcher still crashes without any logs / errors displayed as to what went wrong:
```
22:47:18.699 [Thread-3] TRACE o.t.launcher.LauncherInitTask - Setting up Package Manager
22:47:18.879 [jcabi-loggable] DEBUG com.jcabi.aspects.aj.MethodLogger$2 - Logging monitor thread interrupted
22:47:19.068 [Thread-3] INFO  o.t.l.packages.PackageDatabase - Saved database file: /home/jdr/.terasologylauncher/packages.db
22:47:19.073 [JavaFX Application Thread] ERROR o.t.launcher.TerasologyLauncher - The TerasologyLauncher could not be started!
Exception in thread "JavaFX Application Thread" java.lang.NullPointerException
        at org.terasology.launcher.TerasologyLauncher.openCrashReporterAndExit(TerasologyLauncher.java:150)
        at org.terasology.launcher.TerasologyLauncher.lambda$start$0(TerasologyLauncher.java:125)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:238)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.concurrent.EventHelper.fireEvent(EventHelper.java:219)
        at javafx.concurrent.Task.fireEvent(Task.java:1356)
        at javafx.concurrent.Task.setState(Task.java:707)
        at javafx.concurrent.Task$TaskCallable.lambda$call$81(Task.java:1453)
        at com.sun.javafx.application.PlatformImpl.lambda$null$117(PlatformImpl.java:295)
        at java.security.AccessController.doPrivileged(Native Method)
        at com.sun.javafx.application.PlatformImpl.lambda$runLater$118(PlatformImpl.java:294)
        at com.sun.glass.ui.InvokeLaterDispatcher$Future.run(InvokeLaterDispatcher.java:95)
        at com.sun.glass.ui.gtk.GtkApplication._runLoop(Native Method)
        at com.sun.glass.ui.gtk.GtkApplication.lambda$null$450(GtkApplication.java:139)
        at java.lang.Thread.run(Thread.java:748)
```